File: software/Camera.py
import cv2
import imutils

class Camera:
    def __init__(self,camera_index,fov,frame_width,frame_height,frame_scale_down):
        self.fov = fov
        self.frame_width = frame_width
        self.frame_height = frame_height
        self.scaled_width = frame_width // frame_scale_down
        self.scaled_height = frame_height // frame_scale_down
        self.video = cv2.VideoCapture(camera_index)

    def __del__(self):
        self.video.release()

    def readVideo(self):
        return self.video.read()

def main():
    # Create the Camera object
    camera = Camera(
        camera_index=0,
        fov=70,
        frame_width=1440,
        frame_height=1080,
        frame_scale_down=8
    )

    while True:
        ret, frame = camera.readVideo()
        if not ret:
            break

        # Resize frame for display
        display_frame = imutils.resize(frame, width=camera.scaled_width)

        # Display the resulting frame
        cv2.imshow('Frame', display_frame)

        # Break the loop when 'q' is pressed
        if cv2.waitKey(1) & 0xFF == ord('q'):
            break

    # When everything is done, release the video capture object
    camera.__del__()
    cv2.destroyAllWindows()

if __name__ == "__main__":
    main()

File: rust-rewrite/src/camera.rs
extern crate opencv;

use opencv::prelude::*;
use opencv::videoio::{VideoCapture, VideoCaptureTrait, CAP_ANY};
use std::error::Error;
use opencv::highgui;
use opencv::imgproc;
use opencv::core::Size;
use opencv::core::Mat;

pub struct Camera {
    fov: f32,
    frame_width: i32,
    frame_height: i32,
    scaled_width: i32,
    scaled_height: i32,
    video: VideoCapture,
}

impl Camera {
    pub fn new(camera_index: i32, fov: f32, frame_width: i32, frame_height: i32, frame_scale_down: i32) -> Result<Self, Box<dyn Error>> {
        let video = VideoCapture::new(camera_index, CAP_ANY)?;
        if !video.is_opened()? {
            return Err(From::from("Failed to open camera"));
        }
        let scaled_width = frame_width / frame_scale_down;
        let scaled_height = frame_height / frame_scale_down;
        Ok(Self {
            fov,
            frame_width,
            frame_height,
            scaled_width,
            scaled_height,
            video,
        })
    }

    pub fn read_video(&mut self) -> Result<opencv::core::Mat, Box<dyn Error>> {
        let mut frame = Mat::default();
        self.video.read(&mut frame)?;
        if frame.empty() {
            return Err(From::from("Failed to read frame"));
        }
        Ok(frame)
    }
}

impl Drop for Camera {
    fn drop(&mut self) {
        self.video.release().unwrap();
    }
}

pub fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Create the Camera object
    let mut camera = Camera::new(0, 70.0, 1440, 1080, 8)?;

    loop {
        let frame = camera.read_video()?;
        // Resize frame for display
        let mut display_frame = Mat::default();
        imgproc::resize(&frame, &mut display_frame, Size::new(camera.scaled_width, camera.scaled_height), 0.0, 0.0, imgproc::INTER_LINEAR)?;

        // Display the resulting frame
        highgui::imshow("Frame", &display_frame)?;

        // Break the loop when 'q' is pressed
        if highgui::wait_key(1)? == 'q' as i32 {
            break;
        }
    }
    Ok(())
}
